tests/plugin/test_matcher.py

```diff
@@ -0,0 +1,225 @@
+[pytest]
+allure_labels =
+    """
+    for label in custom_labels:
+        inifile += f"\t{label}\n"
+    pytester.makefile(".ini", pytest=inifile)
+
+
+@allure.step("Check that case has no labels")
+def _has_no_labels(case: dict, labels: Collection[str]) -> None:
+    assert not [label["value"] for label in case["labels"] if label["name"] in labels]
+
+
+@allure.step("Check that case has labels")
+def _has_labels(case: dict, labels: Collection[str], values: Collection[str]) -> None:
+    for label in make_allure_labels(labels, values):
+        assert dict(name=label.name, value=label.value) in case["labels"]
+
+
+@allure.step("Check that entry has link")
+def _has_link(case: dict, needle: str) -> None:
+    assert needle in [link["url"] for link in case["links"]]
+
+
+@allure.step("Find entry {name}")
+def _pop_result(results: MutableSequence[dict], name: str) -> Optional[dict]:
+    for index, value in enumerate(results):
+        if value["name"] == name:
+            return results.pop(index)
+    return None
+
+
+def _has_result(results: MutableSequence[dict], name: str) -> dict:
+    result = _pop_result(results, name)
+    assert result, f"Case with name {name} not found"
+    return result
+
+
+@pytest.mark.usefixtures("_conftest", "_pytestini")
+def test_matcher(
+    pytester: Pytester,
+    test_cases: Tuple[Collection[TestCase], Collection[Scenario]],
+    scenarios: Collection[Scenario],
+    custom_labels: Collection[str],
+):
+    """Test matcher"""
+
+    cases, not_implemented = test_cases
+    pytester_result, allure_results = run_with_allure(
+        pytester=pytester,
+        testfile_path="test_matcher.py",
+        additional_opts=["--sc-type=test"],
+        outcomes=dict(passed=len(cases)),
+    )
+
+    test_items = list(cases)
+    fake_items = [
+        TestCase(
+            name=scenario.display_name,
+            matches=[scenario],
+        )
+        for scenario in not_implemented
+    ]
+
+    with allure.step("Check actual test cases"):
+        for item in test_items:
+            result = _has_result(allure_results.test_cases, item.name)
+            for match in item.matches:
+                _has_link(result, match.link)
+                _has_labels(result, custom_labels, match.specifications_names)
+            if not item.matches:  # check that lables does not applied by mistake
+                _has_no_labels(result, custom_labels)
+
+    with allure.step("Check fake test cases"):
+        for item in fake_items:
+            result = _has_result(allure_results.test_cases, item.name)
+            assert result["status"] == "unknown", "Unexpected entry status"
+            for match in item.matches:
+                _has_link(result, match.link)
+                _has_labels(result, custom_labels, match.specifications_names)
+                _has_labels(result, ScenariosMatcher.DEFAULT_LABELS, match.suites_names)
+
+    assert not allure_results.test_cases, "Unexpected test cases reported"
```

Comment:
I think it is not `AssertionError`. This arises when a developer make mistake (
Added some test cases, but did not process them here), right?

Reply:
The main purpose of this statement is to check that we have only expected test runs in report.
This assert will fail in the case of developer mistake and it's okay, it is still assert and should raise AssertionError.